### PR TITLE
Consume cluster-up from kubevirtci using git subtree

### PR DESCRIPTION
**Pull request by MoserMichael**
**What this PR does / why we need it:**

try to use CI scripts from kubevirt.
pull kubevirt ci via git subtree command.

**Special notes for your reviewer:**

**Release note:**
```
NONE
```

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by beekhof** [comment]
The PR itself records which branch the changes are against, use the title to provide meaningful information

File: cluster-provision [hunk 1]
```diff
@@ -0,0 +1 @@
+vendor/kubevirtci/cluster-provision
```
Comment (beekhof) [inline cluster-provision]: We don't appear to need the `cluster-provision` directory, it would be better not to include it in the first place

Comment (beekhof) [reply cluster-provision]: Actually it would be better to include just the parts we want (pack8s and cluster-up) at the top-level, without any symlinks.  That avoids potential issues with go packaging tools that expect to manage the contents of the `vendor` directory.

File: hack/functest.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+function new_test() {
+    name=$1
+
+    printf "%0.s=" {1..80}
+    echo
+    echo ${name}
+}
+
+new_test 'Test e2e Node Mainenance'
```
Comment (beekhof) [inline hack/functest.sh]: Please add `funktest.sh` and `sync.sh` in the same commit that removes the old `cluster` directory, so that `git` can correctly work out that the files are being moved.
Otherwise there is no way to follow the history of this file - as it appears to come from nowhere in this commit

Comment (MoserMichael) [reply hack/functest.sh]: done.

File: cluster-provision (hunk 1 above)
Comment (MoserMichael) [reply cluster-provision]: moved the kubevirt ci out of vendor, had to force push the whole sequence of commits. Please review again.

File: Makefile [hunk 3]
```diff
@@ -7,6 +7,7 @@ OPERATOR_IMAGE ?= node-maintenance-operator
 REGISTRY_IMAGE ?= node-maintenance-operator-registry
 
 TARGETS = \
+	cluster-up \
```
Comment (beekhof) [inline Makefile]: I assume `pack8s` should also be on this list?

File: cluster-up [hunk 4]
```diff
@@ -0,0 +1 @@
+kubevirtci/cluster-up
```
Comment (beekhof) [inline cluster-up]: If we must have an intermediate `kubevirtci` directory, we should probably just point the `Makefile` at that location rather than create symlinks

File: kubevirtci/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# Contributing to kubevirt/kubevirtci
```
Comment (beekhof) [inline kubevirtci/CONTRIBUTING.md]: Do we not have the option to consume just the parts of `kubevirtci` that we're interested in?
We don't need all this stuff

File: Makefile (hunk 3 above)
Comment (beekhof) [inline Makefile]: What about creating a target for refreshing the subtree?

File: cluster-up (hunk 4 above)
Comment (MoserMichael) [reply cluster-up]: pushed a change to remove the symlinks.

File: Makefile (hunk 3 above)
Comment (MoserMichael) [reply Makefile]: removed the symlink from the top level, so there is now no need to add it to the list.

File: kubevirtci/CONTRIBUTING.md (hunk 5 above)
Comment (MoserMichael) [reply kubevirtci/CONTRIBUTING.md]: I think we should rather have one subtree then a bunch of them.

File: Makefile (hunk 3 above)
Comment (MoserMichael) [reply Makefile]: add pull-ci-changes target to Makefile.

File: hack/sync.sh [hunk 6]
```diff
@@ -1,6 +1,23 @@
...
-source ./cluster-up/hack/common.sh
+if [ -z "$KUBEVIRTCI_PATH" ]; then
+    KUBEVIRTCI_PATH="$(
+        cd "$(dirname "$BASH_SOURCE[0]")/"
+        readlink -f ../kubevirtci/cluster-up
+    )"
+fi
+
+
+if [ -z "$KUBEVIRTCI_CONFIG_PATH" ]; then
```
Comment (beekhof) [inline hack/sync.sh]: `KUBEVIRTCI_CONFIG_PATH` is used by the scripts in `KUBEVIRTCI_PATH`?

Comment (MoserMichael) [reply hack/sync.sh]: added comments to explain the environment variables.

File: hack/functest.sh [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    # sets the path of the CI scripts
+    KUBEVIRTCI_PATH="$(
+        cd "$(dirname "$BASH_SOURCE[0]")/"
+        readlink -f ../kubevirtci/cluster-up
+    )"
+fi
+
+
+if [ -z "$KUBEVIRTCI_CONFIG_PATH" ]; then
+    # set location of ci configuration.
```
Comment (beekhof) [inline hack/functest.sh]: What code does is usually easy enough to figure out, its the *why* that is important to document in situations like this (where you are defining a variable that based on reading the file appears to have no reason to exist).

Comment (MoserMichael) [reply hack/functest.sh]: In

**Comment by yanirq** [comment]
I'm wondering how this is merged eventually and updated .Will we see all of these files of KubeVirt CI merged or as a link ? What happens if something is updated on KubeVirt CI side ? (I would be happy to get a quick process review here)   
Have you considered using [gitmodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) ? (Maybe its here and i missed it)
See example https://github.com/MarSik/kubevirt-ssp-operator/blob/master/.gitmodules

**Comment by MoserMichael** [comment]
> I'm wondering how this is merged eventually and updated .Will we see all of these files of KubeVirt CI merged or as a link ? What happens if something is updated on KubeVirt CI side ? (I would be happy to get a quick process review here)
> Have you considered using [gitmodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) ? (Maybe its here and i missed it)
> See example https://github.com/MarSik/kubevirt-ssp-operator/blob/master/.gitmodules

I can squash my changes into a single commit; however addition of subtree introduces a merge commit with the CI repo.
In any event you could merge this PR as squash.

**Comment by MoserMichael** [comment]
> > I'm wondering how this is merged eventually and updated .Will we see all of these files of KubeVirt CI merged or as a link ? What happens if something is updated on KubeVirt CI side ? (I would be happy to get a quick process review here)
> > Have you considered using [gitmodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) ? (Maybe its here and i missed it)
> > See example https://github.com/MarSik/kubevirt-ssp-operator/blob/master/.gitmodules
> 
> I can squash my changes into a single commit; however addition of a git subtree introduces a merge commit with the CI repo.
> In any event you could merge this PR as squash; so that I don't quite understand your doubts on this issue.

Hi Yanir, In PR #61 you explicitly requested me to do this change based on subtrees, as well 
https://github.com/kubevirt/node-maintenance-operator/pull/61#issuecomment-602187318  as well as Andrew, who requested to do this PR as subtrees https://github.com/kubevirt/node-maintenance-operator/pull/61#issuecomment-599831981 

I understood that you all agreed on the course of action. Please explain you change of mind.

**Comment by yanirq** [comment]
> > > I'm wondering how this is merged eventually and updated .Will we see all of these files of KubeVirt CI merged or as a link ? What happens if something is updated on KubeVirt CI side ? (I would be happy to get a quick process review here)
> > > Have you considered using [gitmodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) ? (Maybe its here and i missed it)
> > > See example https://github.com/MarSik/kubevirt-ssp-operator/blob/master/.gitmodules
> > 
> > 
> > I can squash my changes into a single commit; however addition of a git subtree introduces a merge commit with the CI repo.
> > In any event you could merge this PR as squash; so that I don't quite understand your doubts on this issue.
> 
> Hi Yanir, In PR #61 you explicitly requested me to do this change based on git subtrees, link: [#61 (comment)](https://github.com/kubevirt/node-maintenance-operator/pull/61#issuecomment-602187318) , as well
> as well as Andrew, who requested to do this PR as git subtrees [#61 (comment)](https://github.com/kubevirt/node-maintenance-operator/pull/61#issuecomment-599831981)
> 
> I understood that you all agreed on the course of action. Please explain your change of mind, and why you can't incorporate this PR as a squash commit.
> 
> In addition to that: you can now incorporate future changes of the upstream CI by means of a new target to the makefile: make pull-ci-changes

The suggestions above are only suggestions/queries to see we are not missing something. I'm good with this course of actions as long as we have a stable CI and cleaner code.

**Comment by beekhof** [comment]
> I'm wondering how this is merged eventually and updated .Will we see all of these files of KubeVirt CI merged or as a link ? What happens if something is updated on KubeVirt CI side ? (I would be happy to get a quick process review here)

@MoserMichael, can you please answer these questions for Yanir.

> Have you considered using [gitmodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) ? (Maybe its here and i missed it)
> See example https://github.com/MarSik/kubevirt-ssp-operator/blob/master/.gitmodules

Very similar concepts.  
https://codewinsarguments.co/2016/05/01/git-submodules-vs-git-subtrees/ has a decent overview.

tl;dr
> Submodules are easier to push but harder to pull – This is because they are pointers to the original repository
> Subtrees are easier to pull but harder to push – This is because they are copies of the original repository

My inclination is that if we're making changes to the `kubevirtci` directory directly, something is seriously wrong.  We're less likely to find problems before the maintainers, and we should resolve them upstream even when we do.

**Comment by MoserMichael** [comment]
> > I'm wondering how this is merged eventually and updated .Will we see all of these files of KubeVirt CI merged or as a link ? What happens if something is updated on KubeVirt CI side ? (I would be happy to get a quick process review here)
> 
> @MoserMichael, can you please answer these questions for Yanir.
> 
> > Have you considered using [gitmodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) ? (Maybe its here and i missed it)
> > See example https://github.com/MarSik/kubevirt-ssp-operator/blob/master/.gitmodules
> 
> Very similar concepts.
> https://codewinsarguments.co/2016/05/01/git-submodules-vs-git-subtrees/ has a decent overview.
> 
> tl;dr
> 
> > Submodules are easier to push but harder to pull – This is because they are pointers to the original repository
> > Subtrees are easier to pull but harder to push – This is because they are copies of the original repository
> 
> My inclination is that if we're making changes to the `kubevirtci` directory directly, something is seriously wrong. We're less likely to find problems before the maintainers, and we should resolve them upstream even when we do.

with submodules the travis build script would need to do git fetch after running git clone. I don't know how to customize that step of the build process, the scripts in this repository do not run git clone.

**Comment by fromanirh** [comment]
> > I'm wondering how this is merged eventually and updated .Will we see all of these files of KubeVirt CI merged or as a link ? What happens if something is updated on KubeVirt CI side ? (I would be happy to get a quick process review here)
> 
> @MoserMichael, can you please answer these questions for Yanir.
> 
> > Have you considered using [gitmodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) ? (Maybe its here and i missed it)
> > See example https://github.com/MarSik/kubevirt-ssp-operator/blob/master/.gitmodules
> 
> Very similar concepts.
> https://codewinsarguments.co/2016/05/01/git-submodules-vs-git-subtrees/ has a decent overview.
> 
> tl;dr
> 
> > Submodules are easier to push but harder to pull – This is because they are pointers to the original repository
> > Subtrees are easier to pull but harder to push – This is because they are copies of the original repository
> 
> My inclination is that if we're making changes to the `kubevirtci` directory directly, something is seriously wrong. We're less likely to find problems before the maintainers, and we should resolve them upstream even when we do.
> 
> So for us pull is more important, which suggests subtrees.

My 2c: when I worked on kubevirt (SSP team), we mostly used submodules for two reasons
1. we (or at least I) didn't know about subtrees :)
2. submoduls allowed us to actually keep a strong separation between the code from the project and the imported code. This was actually a side effect that we (or at least I) learned to appreciate after the fact.

We had our fair share of headaches related to submodules, and we ended up keeping our private fork of traviskube scripts: https://github.com/fromanirh/traviskube/ which is somethingwe didn't really want, even though in the end it was less painful than expected.

Worth to be noted that *we* found bugs in the imported traviskube scripts we had to fix, but the only sensible flow was the usual
1. find the bug in the d/s project (e.g. kubevirt-ssp-operator)
2. fix the bug in the u/s project (e.g. traviskube *fork*)
3. pull the changes again (e.g. kubevirt-ssp-operator updates its submodule of traviskube).
Under no circumstances the imported code should be fixed only in the importing repo, as it was already pointed out above.

That said, it seems to me that subtrees are a more convenient alternative for this case provided  that the imported code is updated rarely, if ever.

**Comment by yanirq** [comment]
To sum up the discussion above. continue with subtree and make sure the CI is stable at this point.
@beekhof  see if your requests for change were fulfilled.

**Comment by MoserMichael** [comment]
> To sum up the discussion above. continue with subtree and make sure the CI is stable at this point.
> @beekhof see if your requests for change were fulfilled.

everyone is fine with the approach and the CI is running without problems. At this moment I can' discern any open issues on the implementation of this PR, please correct me if I am wrong.

File: hack/sync.sh (hunk 6 above)
Comment (beekhof) [reply hack/sync.sh:12]: I don't see the changes we discussed on Monday.
Specifically the comment that explains _why_ `KUBEVIRTCI_CONFIG_PATH` needs to be set in a file that does not make use of it.

**Comment by beekhof** [comment]
> > To sum up the discussion above. continue with subtree and make sure the CI is stable at this point.
> > @beekhof see if your requests for change were fulfilled.
> 
> everyone is fine with the approach and the CI is running without problems. At this moment I can' discern any open issues on the implementation of this PR, please correct me if I am wrong.

The changes that document, in the scripts themselves, **_why_** `KUBEVIRTCI_CONFIG_PATH` is being set.

**Comment by beekhof** [comment]
@yanirq I think we're good to go on this now. can we get an approve please?
